### DISCUSSION TITLE
How do I close a socket correctly?

### DISCUSSION THREAD

**Discussion by sishari** [micropython/micropython#13625] [RP2040 / Pico]
I have with pico a socket problem.

A socket server is running on the pico.
I can connect with a client a maximum of 4 times.
After that, no connection with a client is possible. 
How do I properly close a socket on the Pico?

I tried both:
1.
asyncio.start_server(serve, "0.0.0.0", 10001)

async def serve(reader, writer):
     await asyncio.sleep(3)
     writer.close()
     writer.wait_closed()

2.
server1 = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server1.bind(('0.0.0.0', 10001))
client,addr = server1.accept()
await asyncio.sleep(3)
client.close()

**Comment by sosi-deadeye** [comment]
`writer.wait_closed()` is a **coroutine** and you must await for it to finish.
In the second example, you're mixing synchronous code with async code.

Minimal example which should work:


```
import asyncio


async def new_connection(sr: asyncio.StreamReader, sw: asyncio.StreamWriter):
    addr = sr.get_extra_info("peername")
    print("Got new connection:", addr)
    print("Closing connection:", addr)
    
    data = await sr.read()
    sw.write(b"Re: " + data)
    await sw.drain()
    
    sw.close()
    await sw.wait_closed()


async def main():
    server = await asyncio.start_server(new_connection, "0.0.0.0", 8000)
    await server.wait_closed()


if __name__ == "__main__":
    asyncio.run(main())

```

**Reply by sishari** [reply]
Hello,

I modified your code and it worked.
But I still have a problem:
If a client is connected, I send data every 2 seconds.
How can I detect that a client no longer exists.
e.g.: Network cable unplugged.
At the moment the socket server is still sending.
I would like that the socket server close the open port to the client.
